src/shim/gate.h

```diff
@@ -0,0 +1,25 @@
+#ifndef GATE_H_
+#define GATE_H_
+
+#include <atomic>
+#include <cstddef>
+
+#include <pthread.h>
+#include <semaphore.h>
+
+/*
+ * Intended to be private to IPC module.
+ */
+
+struct Gate {
+  std::atomic<bool> x;
+  sem_t semaphore;
+  std::size_t spin_ctr;
+  pthread_spinlock_t lock;
+};
+
+void gate_init(Gate *gate);
```

Comment:
I'm struggling a little bit with the "gate" metaphor. Could perhaps use a brief explanation (and/or a reference if it's a standard term, though I didn't have much luck searching)

Reply:
Renamed to BinarySpinningSem